src/synthetic_dataset_generator/apps/rag.py

```diff
@@ -0,0 +1,900 @@
+    progress(
+        1,
+        total=total_steps,
+        desc="Creating dataset",
+    )
+
+    # create distiset
+    distiset_results = []
+    source_results = reranking_results if reranking else response_results
+    base_keys = ["context", "question", "response"]
+    retrieval_keys = ["positive_retrieval", "negative_retrieval"] if retrieval else []
+    reranking_keys = ["positive_reranking", "negative_reranking"] if reranking else []
+    relevant_keys = base_keys + retrieval_keys + reranking_keys
+
+    for result in source_results:
+        record = {key: result.get(key) for key in relevant_keys if key in result}
+        distiset_results.append(record)
+
+    dataframe = pd.DataFrame(distiset_results)
+
+    progress(1.0, desc="Dataset generation completed")
+    return dataframe
+
+
+def generate_sample_dataset(
+    repo_id: str,
+    file_paths: list[str],
+    input_type: str,
+    system_prompt: str,
+    document_column: str,
+    retrieval_reranking: list[str],
+    num_rows: str,
+    oauth_token: Union[OAuthToken, None],
+):
+    retrieval = "Retrieval" in retrieval_reranking
+    reranking = "Reranking" in retrieval_reranking
+
+    if input_type == "prompt-input":
+        dataframe = pd.DataFrame(columns=["context", "question", "response"])
+    else:
+        dataframe, _ = load_dataset_file(
+            repo_id=repo_id,
+            file_paths=file_paths,
+            input_type=input_type,
+            num_rows=num_rows,
+            token=oauth_token,
+        )
+    dataframe = generate_dataset(
+        input_type=input_type,
+        dataframe=dataframe,
+        system_prompt=system_prompt,
+        document_column=document_column,
+        retrieval=retrieval,
+        reranking=reranking,
+        num_rows=10,
+        is_sample=True,
+    )
+    return dataframe
+
+
+def push_dataset_to_hub(
+    dataframe: pd.DataFrame,
+    org_name: str,
+    repo_name: str,
+    oauth_token: Union[gr.OAuthToken, None],
+    private: bool,
+    pipeline_code: str,
+):
+    repo_id = validate_push_to_hub(org_name, repo_name)
+    dataset = Dataset.from_pandas(dataframe)
+    dataset = combine_datasets(repo_id, dataset, oauth_token)
+    distiset = Distiset({"default": dataset})
+    distiset.push_to_hub(
+        repo_id=repo_id,
+        private=private,
+        include_script=False,
+        token=oauth_token.token,
+        create_pr=False,
+    )
+    push_pipeline_code_to_hub(pipeline_code, org_name, repo_name, oauth_token)
+
+
+def push_dataset(
+    org_name: str,
+    repo_name: str,
+    private: bool,
+    original_repo_id: str,
+    file_paths: list[str],
+    input_type: str,
+    system_prompt: str,
+    document_column: str,
+    retrieval_reranking: list[str],
+    num_rows: int,
+    temperature: float,
+    pipeline_code: str,
+    oauth_token: Union[gr.OAuthToken, None] = None,
+    progress=gr.Progress(),
+) -> pd.DataFrame:
+    retrieval = "Retrieval" in retrieval_reranking
```

Comment:
we can already add a `process` at the  beginning of each function call to ensure we directly get visual feedback in the app.